scripts/check_compliance.py

```diff
@@ -211,6 +215,124 @@ def run(self):
 """.format(max_top_items, n_top_items, __file__)
 
 
+        #
+        # Look for references to undefined Kconfig symbols outside Kconfig
+        # files
+        #
+
+        # Grep for symbol references.
+        #
+        # Example output line for a reference to CONFIG_BAZ at line 17 of
+        # foo/bar.c:
+        #
+        #   foo/bar.c<null>17<null>CONFIG_BAZ
+        #
+        # Skip the samples/ and tests/ directories for now. They often contain
+        # Kconfig files that are not part of the main Kconfig tree, which will
+        # trigger false positives until we do something fancier. Skip
+        # doc/releases too, which often references removed symbols.
+        grep_cmd = "git grep --only-matching --line-number -I --null " \
+                   "--extended-regexp --word-regexp CONFIG_[A-Z0-9_]+ " \
+                   "-- :!samples :!tests :!doc/releases"
+
+        grep_process = subprocess.Popen(grep_cmd.split(),
+                                        stdout=subprocess.PIPE,
+                                        stderr=subprocess.PIPE,
+                                        cwd=self.zephyr_base)
+
+        grep_stdout, grep_stderr = grep_process.communicate()
+        if grep_process.returncode:
+            self.case.result = Error("error running 'git grep'", "error")
+            # grep_stderr will be 'bytes' here, which prints as b'oops'
+            self.case.result._elem.text = \
+                "'{}' failed with exit code {} and wrote {} to stderr " \
+                "(while searching for Kconfig symbol references)" \
+                .format(grep_cmd, grep_process.returncode, grep_stderr)
+            return
+
+        defined_syms = set(sym.name for sym in kconf.unique_defined_syms)
+        undef_to_locs = collections.defaultdict(list)
+
+        # splitlines() supports various line terminators
+        for line in grep_stdout.decode("utf-8").splitlines():
+            path, lineno, sym_name = line.split("\0")
+
+            # [7:] removes the "CONFIG_" prefix
+            if sym_name[7:] not in defined_syms and \
+               sym_name not in UNDEF_KCONFIG_WHITELIST:
+
+               undef_to_locs[sym_name].append("{}:{}".format(path, lineno))
+
+        if undef_to_locs:
+            undef_desc = "\n".join(
+                "{:35} {}".format(sym_name, ", ".join(locs))
```

Comment:
I don't understand what undef_desc is and where this particular sym_name comes from.

Reply:
It's a description of the undefined variables.

The high-level format is this:

```python
undef_desc = "\n".join(... for sym_name, locs in ...)
```

Can try this if you haven't run across generators before:

```python
" foo ".join(2*s for s in ("bar", "baz", "qaz"))
```

Reply:
Added a comment above `undef_desc` in the code.

Reply:
Got it, the `for` keyword at the start of the line (and the late hour...) confused me. How about moving the `for` at the end of the previous line instead? Neither pylint nor pep8 complains.

Reply:
> Suggestions for a better name would be appreciated.

Something like `undef_failure_msg`?

Reply:
It isn't the entire message, just the description of the undefined variables, so might be a bit misleading (unless the code is changed to build up the entire message in that variable, but think that'd be overkilling it).

I added a comment to clarify it though.